storage/log_storage.go

```diff
@@ -100,7 +100,7 @@ type ReadOnlyLogStorage interface {
 	// Commit must be called when the caller is finished with the returned object,
 	// and values read through it should only be propagated if Commit returns
 	// without error.
-	SnapshotForTree(ctx context.Context, treeID int64) (ReadOnlyLogTreeTX, error)
+	SnapshotForTree(ctx context.Context, treeID int64, opts GetOpts) (ReadOnlyLogTreeTX, error)
```

Comment:
Aren't the ops all implicit here (and in anything implementing `storage/map_storage.go` too) according to which method is being called?

Reply:
Not quite sure what you mean. Do you mean we don't have fine grained control on read access at the moment? I wanted to add this to all the APIs now so we can take the hit once in case we do later.

Reply:
Also the underlying problem is that once you get into storage it calls GetTree again, which needs a GetOpts. This is why I'm pushing it all the way down in all the calls.

Reply:
It feels natural to pass `GetOpts` to `SnapshotForTree` and `RWTx` because those methods don't know what kind of transaction follows. I am not sure about `QueueLeaves` though. We know exactly what it does and which `GetOpts` should be inside. If the only reason to pass `GetOpts` to `QueueLeaves` is avoiding copy-paste then we could probably enumerate access modes for operations somewhere next to storage API:

```
var (
  QueueLeavesAccessMode = NewGetOpts(Queue, ReadWrite, trillian.TreeType_LOG)
  LogReadAccessMode = NewGetOpts(Query, ReadOnly, trillian.TreeType_LOG)
  // etc.
)
```

Reply:
It's arguably the wrong place for it though. Policy should probably be handled higher up and this layer shouldn't have to worry about it. Anyway I'm now doing another huge refactor to see how that would look.